Migrate ui/doc_title to New platform

pgayvallet opened the pull request:
## Summary

Migrate ui/doc_title to new platform:
- Add a DocTitle sub-service to ChromeService
- Bridge the legacy `ui/doc_title` to use the NP service

Fixes #46077

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

~~- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~~
~~- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~~
- [X] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [X] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
~~- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~~

### For maintainers

~~- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~
~~- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~~

## Dev Docs

Migrate `chrome.docTitle` to new platform. Plugins can now change the page title using this API.

```ts
coreStart.docTitle.change('My Title');
```

elasticmachine commented:
Pinging @elastic/kibana-platform (Team:Platform)

pgayvallet commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -0,0 +1,180 @@
+  __legacy: {
+    setBaseTitle: (baseTitle: string) => void;
+  };
+}
```
We don't have any other services with that kind of legacy API, however this was needed to keep the old docTitle karma test working, as they were using 'private' APIs to manually change to baseTitle of the old module.  Please tell me if these is a better approach

joshdover replied:
What I've been doing with the legacy tests is just making them dumb unit tests that ensure the NP API was called by using a mock. That way we don't have to do things like this, which should save us time when we start removing the legacy code.

pgayvallet replied:
@joshdover issue I had here is with the tests in `src/legacy/ui/public/doc_title/__tests__/doc_title.js` that were basically injected in the old docTitle module a different base name. Could find another way than than to keep the test. Do you see another approach ? Or maybe should I just remove the legacy test as we now got that covered in the NP module tests instead ?

joshdover replied:
I'm fine with removing this test completely.

pgayvallet commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -0,0 +1,180 @@
+ *
+ * @public
+ */
+export type ChromeDocTitleChange = string | string[] | ChromeDocTitleEntry;
```
This is a only a nice to have, but it felt like this could ease the developer way to call the API. If we prefer something more rigid, we can stick to `ChromeDocTitleEntry` instead

rudolf replied:
Now that I think more about this... this service does just two things, if you give it parts it joins them with ` - ` and it adds the baseTitle. It's so small it's almost annoying :-P but the value it provides is consistency for all of Kibana. If a plugin really needs to break this consistency it feels like they might as well just use `document.title = `.

rudolf replied:
The ingest team is happy for us to drop the `true` flag in this PR so that monitoring also uses the baseTitle.

pgayvallet replied:
@rudolf ok great. then I guess I will change the type to `export type ChromeDocTitleChange = string | string[]`

pgayvallet replied:
But yes, the dual logic between displaying the title and the `join(' - ')` is quite annoying :)

pgayvallet commented on src/legacy/ui/public/doc_title/__tests__/doc_title.js:
```diff
@@ -30,11 +31,11 @@ describe('docTitle Service', function () {
     initialDocTitle = document.title;
     document.title = MAIN_TITLE;
-    setBaseTitle(MAIN_TITLE);
+    npStart.core.chrome.docTitle.__legacy.setBaseTitle(MAIN_TITLE);
```
This is what I'm referring to in previous comment.

pgayvallet commented on src/legacy/ui/public/doc_title/doc_title.js:
```diff
@@ -17,43 +17,28 @@
 }
 
 export const docTitle = {
-  render,
```
render seems to be an unused export, so removing it was avoiding another method bridging.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4362/)
* Commit: 122b7d5fae56d9533c5d291b79740870e754395e

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/4430/)
* Commit: 08f20ade8468ad4599b60a3c88c17eb27e66ed41

joshdover commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -0,0 +1,180 @@
+export class DocTitleService {
+  private document = { title: '' };
+  private baseTitle = '';
+  private current = defaultTitle;
```
Why use an additional property instead of `this.title$.value`? Seems like unnecessary state to track.

pgayvallet replied:
indeed it is, thanks

joshdover commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -0,0 +1,180 @@
+    this.title$.next(this.baseTitle);
+
+    return {
+      get$: () => this.title$.pipe(takeUntil(this.stop$)),
```
What actually uses this?

pgayvallet replied:
No one atm. Just thought it might be useful to expose it so its already present if a plugin may want to use it in the futur. Do you think it's better to just remove it ?

pgayvallet commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -0,0 +1,180 @@
+   * @param newTitle The new title to set, either a string, string array or {@link ChromeDocTitleEntry}
+   * @param [apply=true] If false, will not actually apply the new title until #apply() is called.
+   */
+  change: (newTitle: ChromeDocTitleChange, apply?: boolean) => void;
```
Well, thats pretty good to know.

pgayvallet commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -0,0 +1,180 @@
+   * @param newTitle The new title to set, either a string, string array or {@link ChromeDocTitleEntry}
+   * @param [apply=true] If false, will not actually apply the new title until #apply() is called.
+   */
+  change: (newTitle: ChromeDocTitleChange, apply?: boolean) => void;
```
The legacy code was doing some questionable code with this option yes (basically reset without apply on route change start, then apply on route change success), thats why I was afraid to drop it.

rudolf replied:
yeah I don't think we can get away from reset supporting `apply=false` behaviour without breaking legacy behaviour, but I don't think we need this in the change() method

pgayvallet replied:
actually we might be able to. The two-step reset using angular handlers doesnt seems to be really of any usage. CI will confirm or infirm this, but gonna try to remove it

pgayvallet commented:
I've made the changes.
- `__legacy.setBaseTitle` has been kept,  as without it I had to basically remove every old karma tests
- I removed the `apply` option and logic. It was only used for the angular router events, and was not, in fact, required. Now, we only call `reset` during the route change start event, and the behaviour appears to be exactly the same (and CI seems to confirm this)
- removed the `$get` and `currentTitle` as they were unused.
- removed the `excludeBase` option
- doc updated and entry in migration guide added.

This greatly simplify the code and test

@rudolf and @joshdover PTAL

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6398/)
* Commit: 142814926121e4d776521e3e2ae42271b962b7cf

joshdover commented:
ack: Will re-review today

rudolf commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -43,113 +41,56 @@ interface StartDeps {
  * @public
  */
-export type ChromeDocTitleChange = string | string[] | ChromeDocTitleEntry;
+export type ChromeDocTitleChange = string | string[];
```
optional: Is there any advantage to having a named type? It feels like it might be simpler if to inline if we're only using primitive types, but happy for you to merge this either way.

pgayvallet replied:
I had the same question. I'm quite used to always use named types from were I worked before, but for that kind of scenario, I cant really think of any real advantage (expect maybe refactoring if one day we modify/extend the type), and I agree that I doesn't really makes documentation more clear. If we think this is better I definitely can fall back to inline types

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6751/)
* Commit: e4b236ad8071499fd757e05e64f14de980d01f93
